### ISSUE TITLE
Dramatic performance difference running NeMo encoder comparing Tesla T4 and A100 accelerators

### ISSUE THREAD

**Issue by pks** [NVIDIA/TensorRT#2363]
## Description

While the performance of the encoder of [NeMo's MTEncDec](https://github.com/NVIDIA/NeMo/blob/4cd9b3449cbfedc671348fbabbe8e3a55fbd659d/examples/nlp/machine_translation/enc_dec_nmt.py) model is great on an A100 when using TensorRT, it is much less so on a Tesla T4 where TensorRT underperforms ONNXRUNTIME using the `CUDAExecutionProvider`.
Encoding 1000 random inputs takes about 30 seconds with TensorRT on a T4, 5 seconds with TensorRT on an A100, and with ONNXRUNTIME on a T4 it's 25 seconds.

I would expect TensorRT to outperform ONNXRUNTIME on a T4.

## Environment

**TensorRT Version**: 8.4.2
**NVIDIA GPU**: A100-SXM4-40GB / Tesla T4
**NVIDIA Driver Version**: 470.82.01 (A100) / 515.65.01 (T4)
**CUDA Version**: 11.7
**CUDNN Version**: 8.5.0
**Operating System**: CentOS
**Python Version (if applicable)**: 3.8.10 
**Tensorflow Version (if applicable)**: -
**PyTorch Version (if applicable)**: -
**Baremetal or Container (if so, version)**: container nvcr.io/nvidia/tensorrt:22.08-py3

## Relevant Files

A. Script to export encoder graph from NeMo:
```
from nemo.collections.nlp.models import MTEncDecModel

model = MTEncDecModel.from_pretrained("nmt_en_de_transformer12x2")
model.encoder.export("encoder.onnx")
```

B. ONNXRUNTIME test script:
```
import numpy as np
import onnxruntime
import sys
import time

sess_options = onnxruntime.SessionOptions()
sess_options.graph_optimization_level = onnxruntime.GraphOptimizationLevel.ORT_ENABLE_ALL

providers = [
    ('CUDAExecutionProvider', {
        'device_id': 0,
        'arena_extend_strategy': 'kNextPowerOfTwo',
        'gpu_mem_limit': 2 * 1024 * 1024 * 1024,
        'cudnn_conv_algo_search': 'EXHAUSTIVE',
        'do_copy_in_default_stream': True,
    }),
    'CPUExecutionProvider',
]

encoder_session = onnxruntime.InferenceSession(sys.argv[1], providers=providers, sess_options=sess_options)

iters = 1000
max_len = 512
examples = []
for i in range(iters):
    len_ = np.random.randint(1, 512)
    examples.append((np.random.randint(30000, size=[1, len_], dtype=np.int64),
                     np.ones([1, len_], dtype=np.int64))
                   )

start = time.perf_counter()
for i in range(iters):
    encoder_input = {
        "input_ids": examples[i][0],
        "encoder_mask": examples[i][1],
    }
    encoder_output_names = ["last_hidden_states"]
    _ = encoder_session.run(encoder_output_names, encoder_input)[0]
total_time = time.perf_counter() - start
print(f"{total_time}s")
```

C. Starting TensorRT's docker container for the following steps:
```
nvidia-docker run --gpus all -it --rm nvcr.io/nvidia/tensorrt:22.08-py3
```

D. TensorRT conversion script:
```
#!/usr/bin/env bash

trtexec \
        --workspace=10000 \
        --onnx=encoder.onnx  \
        --saveEngine=encoder.trt-engine \
        --minShapes=input_ids:1x1,encoder_mask:1x1 \
        --maxShapes=input_ids:1x512,encoder_mask:1x512 \
        --optShapes=input_ids:1x512,encoder_mask:1x512
```

E. TensorRT inference w/ polygraphy script:
```
import numpy as np
from polygraphy.backend.common import BytesFromPath
from polygraphy.backend.trt import EngineFromBytes, TrtRunner
import time

iters = 1000
max_len = 512
examples = []
for i in range(iters):
    len_ = np.random.randint(1, max_len)
    examples.append((np.random.randint(30000, size=[1, len_], dtype=np.int32),
                     np.ones([1, len_], dtype=np.int32))
                   )

load_engine = EngineFromBytes(BytesFromPath("encoder.trt-engine"))
with TrtRunner(load_engine) as runner:
    start = time.perf_counter()
    for i in range(iters):
        _ = runner.infer({"input_ids":examples[i][0],
                          "encoder_mask":examples[i][1]})
    end = time.perf_counter()
total_time = end - start
print(f"{total_time}s")
```

## Steps To Reproduce

1. Export `encoder.onnx` using script A
2. Run inference w/ ONNXRUNTIME using script B
3. Start docker container using script C
4. Convert `encoder.onnx` to TensorRT engine using script D
5. Run inference w/ polygraphy using script E

Run steps 3-5 with either T4 or A100 GPUs. Observe performance differences.

**Comment by zerollzeng** [comment]
I've file internal bug 3823757 for this.

cc @nvpohanh

**Comment by nvpohanh** [comment]
I can replicate the results and it is because TRT 8.4 has sub-optimal support for dynamic shapes for Transformers. In TRT 8.5 GA that will be published soon, there will be a new preview feature flag that will allow you to get a TRT engine with better dynamic shape support. Here are the performance numbers I got on the T4 machine I have:

- onnx-runtime: 22.91s
- TRT 8.4: 28.01s
- TRT 8.5: 27.06s
- TRT 8.5 + preview flag enabled (i.e. build engine with `--preview=+fasterDynamicShapes0805`): 15.58s

**Comment by nvpohanh** [comment]
this is because for ONNXRT you are using random SeqLen between [1, 512] while for TRT, you are using fixed SeqLen=512 so it's not apples-to-apples comparison.

From offline discussion, this will be further improved if you use TRT 8.5 with kFASTER_DYNAMIC_SHAPES_0805 preview flag. 

Closing this.

**Comment by pks** [comment]
> this is because for ONNXRT you are using random SeqLen between [1, 512] while for TRT, you are using fixed SeqLen=512 so it's not apples-to-apples comparison.

Could you point out my error where I don't make use of dynamic shapes? The examples in the python code are generated the same way between ORT and TRT and I thought I'd make use of dynamic shapes via these
```
        --minShapes=input_ids:1x1,encoder_mask:1x1 \
        --maxShapes=input_ids:1x512,encoder_mask:1x512 \
        --optShapes=input_ids:1x512,encoder_mask:1x512
```
arguments to `trtexec`.

**Comment by nvpohanh** [comment]
Yes, that trtexec command generates an engine that can run any SeqLen between 1 and 512, but when trtexec runs the inference, it use the shapes provided with `--shapes` flag to run the inference. If `--shapes` is not provided, it uses `--optShapes` to run the inference. So effectively, you are measuring TRT's perf with SeqLen=512. Currently, trtexec does NOT support running with random input shapes, so you will need to write your own python harness to measure TRT perf.

**Comment by pks** [comment]
> but when trtexec runs the inference

I'm using polygraphy to run the graph, wouldn't that work?

**Comment by nvpohanh** [comment]
@pranavm-nvidia Does Polygraphy support random input shapes? I don't recall that we have added that support

**Comment by pranavm-nvidia** [comment]
@nvpohanh The command-line tool doesn't randomize the input shape but in this case it looks like the benchmark is being done manually via a Python script (see part (E) in the description). 

@pks Polygraphy is primarily intended for debugging and fast prototyping and is not reliable for perf measurements as it does introduce some overheads compared to e.g. `trtexec`. Some of that is related to one-time initialization costs, which you could eliminate by doing some warm-up runs before you start timing, but even then there would still be non-negligible overhead.

**Comment by nvpohanh** [comment]
Ah sorry, I totally misunderstood the commands. 

Yes, the step (E) is the **correct** way to measure TRT perf with dynamic input shapes. And now I recall the correct fix: @pks when you are building the engine in step (D), could you add `--preview=+fasterDynamicShapes0805` flag? This enables a preview feature which speeds up the Transformer perf when input shapes are dynamic. Thanks

**Comment by ttyio** [comment]
closing since no activity for more than 3 weeks, thanks!
